### PR TITLE
hls-eval-plugin: Replicate #4139

### PR DESCRIPTION
**Pull request by mattapet**

### CODE REVIEW

**Comment by fendor** [comment]
Thanks! As a note, our CI jobs are rather flaky and fail unpredictably, so don't be afraid and suspect an error in your code if some random plugin tests fail.

**Comment by fendor** [comment]
That's a good start! Perhaps now we can investigate why the other test cases are fine, and only this scenario is affected?

Since it might not be obvious, you can run your individual test case locally like this:

```sh
> TASTY_PATTERN="T4139" cabal test test:hls-eval-plugin-tests
```

To see debug logs:
```sh
> HLS_TEST_PLUGIN_LOG_STDERR=1 TASTY_PATTERN="T4139" cabal test test:hls-eval-plugin-tests
```

And to log *everything* in HLS (likely too much information):

```sh
> HLS_TEST_LOG_STDERR=1 TASTY_PATTERN="T4139" cabal test test:hls-eval-plugin-tests
```

**Comment by mattapet** [comment]
The debug output was not much of a help. In continuing the search, I tried explicitly filtering out the `main` function from the renamed environment, which made the tests pass again (at least locally).

@fendor do you know if there are any specific properties of the module, or the hsc environment that it might be worth to check for differences?

**Comment by fendor** [comment]
> do you know if there are any specific properties of the module, or the hsc environment that it might be worth to check for differences?

Unfortunately no, I am not familiar with that part of GHC's codebase. Perhaps @wz1000 knows.

**Comment by mattapet** [comment]
Continuing the search, I found the [tcRnDeclsi](https://hackage.haskell.org/package/ghc-9.8.2/docs/src/GHC.Tc.Module.html#tcRnDeclsi) in [hscParsedDecls](https://hackage.haskell.org/package/ghc-9.8.2/docs/src/GHC.Driver.Main.html#hscParsedDecls) to be the point where the `evalPrint` is last seen in the environment. After switching the `ddump-rn-trace` I also noticed the difference in the exports there:

`rnExports: Exports: [main]` vs `rnExports: Exports: [evalPrint]` when we filter out the `main` symbol from the initial env. I'll check if traces further for additional leads. I can share them if that would help

**Comment by fendor** [comment]
Sounds like good detective work! You can share whatever you find!

I will likely only get to it some time in the next couple of days.

**Comment by mattapet** [comment]
Attaching the diffs. They're produced by running `cabal test test:hls-eval-plugin-tests > <<dump-file.txt>>`. I'll also update the PR with version adding the dump flag itself in a bit.

[rn-dump-main.txt](https://github.com/haskell/haskell-language-server/files/14593149/rn-dump-main.txt)
[rn-dump-no-main.txt](https://github.com/haskell/haskell-language-server/files/14593150/rn-dump-no-main.txt)

**Comment by mattapet** [comment]
Scratch the previous comment. It turned out to be a dead-end. After comparing with renamer traces with GHC `9.4.8`, which is the version where the test works, I found the same `Exports`. But I decided to add couple more dump flags and compare the traces across the `9.4.8` and `9.8.2` without changes, and `9.8.2` with `main` filtering.

It appears that for `9.8.2` with the `main` function present, the `evalPrint` function gets removed between `ddump-ds-preopt` and `ddump-ds` steps. Adding traces files here:

[output-dump-9.4.8.txt](https://github.com/haskell/haskell-language-server/files/14593733/output-dump-9.4.8.txt)
[output-dump-9.8.2.txt](https://github.com/haskell/haskell-language-server/files/14593734/output-dump-9.8.2.txt)
[output-dump-no-main-9.8.2.txt](https://github.com/haskell/haskell-language-server/files/14593735/output-dump-no-main-9.8.2.txt)

**Comment by michaelpj** [comment]
Can we merge the test as an expected failure for now? It's a good contribution regardless!

File: plugins/hls-eval-plugin/test/Main.hs [hunk 1]
```diff
@@ -88,6 +88,7 @@ tests =
   , goldenWithEval "Reports an error for an incorrect type with :kind" "T13" "hs"
   , goldenWithEval' "Returns a fully-instantiated type for :type" "T14" "hs" (if ghcVersion >= GHC98 then "ghc98.expected" else "expected") -- See https://gitlab.haskell.org/ghc/ghc/-/issues/24069
   , knownBrokenForGhcVersions [GHC92, GHC94, GHC96, GHC98] "type +v does not work anymore with 9.2" $ goldenWithEval "Returns an uninstantiated type for :type +v, admitting multiple whitespaces around arguments" "T15" "hs"
+  , goldenWithEval "T4139" "T4139" "hs"
```
Comment (jhrcek) [inline plugins/hls-eval-plugin/test/Main.hs:91]: Looks good.
Now just mark it as known broken (knownBrokenForGhcVersions [GHC98] $ goldenWithEval ...
and give it more human readable test name, like "Doesn't break in module containing main function".

Comment (jhrcek) [inline plugins/hls-eval-plugin/test/Main.hs:91]: Looks good.
Now just mark it as known broken and give it more human readable test name, like:

```suggestion
  , knownBrokenForGhcVersions [GHC98] $ goldenWithEval "Doesn't break in module containing main function" "T4139" "hs"
```

**Comment by soulomoon** [comment]
It might worth add a eval benchmark test to see if this would cause any leak or undesired behaviour.

**Comment by mattapet** [comment]
@soulomoon I added a couple of experiments, one with very simple, single line evaluation, and one a bit more complex. After inspecting the output logs locally, the benchmarks seem to actually exercise the desired behavior. Is this what you had in mind though?

**Comment by soulomoon** [comment]
> @soulomoon I added a couple of experiments, one with very simple, single line evaluation, and one a bit more complex. After inspecting the output logs locally, the benchmarks seem to actually exercise the desired behavior. Is this what you had in mind though?

Thank you for add new benchmarks, it is looking good.
If we are not seeing any more memory usage etc,. It should be good to go.

**Review by soulomoon — APPROVED** [review]
LGTM

**Pull request merged**